### ISSUE TITLE
Controller can't handle more than 2 GET methods

### ISSUE THREAD

**Issue by EddyRubens** [tomkuijsten/restup#57]
I have a Controller that has 3 GET methods. Each of those methods has a different number of parameters. Here are the UriFormat definitions of the methods:

[UriFormat("/api/collections/{collectionName}/artists/{artistName}")]
[UriFormat("/api/collections/{collectionName}")]
[UriFormat("/api/collections")]

The first method never gets called. Is this 'by design'?

**Comment by tomkuijsten** [comment]
Kindda 😃 I don't think we want to allow a difference between a Uri with our without a trailing slash. Doesn't feel right from REST perspective. The Uri should be clear about its functionality. So the better question is, why two Uri's that are so similar and have the same verb?

**Comment by Jark** [comment]
I think what @EddyRubens is highlighting is that given the following code:

```
public class SomeController
{
    [UriFormat("/collections/{collectionName}/artists/{artistName}")]
    public IGetResponse Get(string collectionName, string artistName)
    {
        return new GetResponse(GetResponse.ResponseStatus.OK);
    }

    [UriFormat("/collections/{collectionName}")]
    public IGetResponse Get(string collectionName)
    {
        return new GetResponse(GetResponse.ResponseStatus.OK);
    }

    [UriFormat("/collections")]
    public IGetResponse Get()
    {
        return new GetResponse(GetResponse.ResponseStatus.OK);
    }            
}
```

He would expect all three calls to be mapped correctly based on how specific your url is.

Imho it would be nice to support this, but if we explicitly don't want to support this than we should document this somewhere and add a unit test to prove this doesn't work.

**Comment by EddyRubens** [comment]
@Jark you are right, this what I meant.
I was converting a web api service where this possible.
I had a look at the code there is an exception on a 'restMethods.SingleOrDefault' in RestControllerRequestHandler that has too many matches. I'll check later if I can workaround...

**Comment by tomkuijsten** [comment]
Oh sorry, didn't see the /{collectionName} (mobile website formatting I guess. That should work ofcourse. Probably a problem in the matching algorithm. I'll fix it when I'm home (away for the weekend). Or maybe @Jark can have a look.

**Comment by EddyRubens** [comment]
There is a relatively easy fix. As all methods are working well individually, it is just a matter of picking just one of the matches. I propose to pick the one with most parameters.

For this, only one line of code needs to be changed:

The following line in RestControllerRequestHandler.cs:
var restMethod = restMethods.SingleOrDefault(r => r.Verb == req.HttpServerRequest.Method);

Could be replaced by:
var restMethod = restMethods.OrderByDescending(r => r.MethodInfo.GetParameters().Count()).FirstOrDefault();

That fixed the issue for me. I know that a complete sort of the methods is overkill, and having seen your coding, I am convinced you will come up with something way more elegant and faster 😃.

Thanks for your quick response, have a nice weekend!

**Comment by Jark** [comment]
Hey @EddyRubens,

That's a lot simpler solution then I had in mind! Thanks for the idea.

I think the design should be:
- do the sorting when the methods are being added, e.g. use a `SortedSet` like for the RouteRegistrations in `HttpServer` 
- make `RestControllerMethodInfo` implement `IComparable` or use a separate comparer
- change the restMethods to do a LastOrDefault / FirstOrDefault depending on the sort order

I don't mind doing this. I'll assign it to myself.

Cheers,

Jark

**Comment by EddyRubens** [comment]
Hi @Jark,

Thanks for taking it up. Your solution sounds right. The problem with my solution is that I do a sort while only a taking the maximum is needed. I had to because there is no MaxBy on the collection and I wanted to keep a one line solution.
Because you build a SortedSet from the start, the effort for getting a sorted list is optimal!

**Comment by tomkuijsten** [comment]
Fixed with pr #58
